articles/create-a-node-cli/index.md

```diff
@@ -0,0 +1,265 @@
+const chalk = require("chalk");
+
+let url = "";
+
+function getJoke() {
+  url = "https://official-joke-api.appspot.com/random_joke";
+ axios({
+ method: "get",
+ url: url,
+ })
+ .then((res) => {
+ const setup = chalk.cyan(res.data.setup);
+ const punchline = chalk.green(res.data.punchline);
+ console.log(`${setup} - ${punchline}`);
+ })
+ .catch((err) => {
+ const log = chalk.red(err);
+ console.log(log);
+ });
+}
+
+function getQuote() {
+axios({
+    method: 'get',
+    url: "http://quotes.stormconsultancy.co.uk/random.json",
+}).then((res)=> {
+    const randomQuote = chalk.cyan(res.data.quote);
+    const randomAuthor = chalk.green(res.data.author);
+    const log = `${randomQuote} - ${randomAuthor}`;
+    console.log(log);
+}).catch((err)=>{
+    const errorLog = chalk.red(err);
+    console.log(errorLog);
+})
+```
+
+Note: We are using chalk to color our logs and get a colorful output. We setup an empty string called `url` so that when the user selects quotes, we load in the quote API URL and the same goes for jokes.
+
+For now, the code won't work because we're not calling any of the functions.
+
+## Understanding option/command/argument parsing
+
+A good example of option parsing is when we passed _-y_ to our _npm init_ command to set everything to default.
+
+The commands used in CLI applications often come with other options or flags. They extend the functionality of the application like in the example above.
+
+A common application of options is when asking for user input which determines the output of the application.
+
+Node.js comes with an in-built app interface for the process module called `process.argv`. It takes in the arguments passed, invokes a method and returns an array with all the arguments passed into the main process.
+
+Using _process.arg_ might be tedious and not beginner friendly so we will use [yargs](https://www.npmjs.com/package/yargs) to help us process the options.
+
+Yargs offers a lot of options that make it very powerful and widely used. Some of them include `.command()`, `.help()`, `.example()`, `.usage()` and others. To read more on the different capabilities of the options see the yargs [docs](https://github.com/yargs/yargs).
+
+We will be making use of the command option to allow us to invoke a function when the set command is triggered. The syntax for this is:
+
+```javascript
+const argv = require("yargs");
+argv.command(
+  "name of command",
+  "description of the command",
+  functionA,
+  functionB
+).argv;
+```
+
+Note: _functionA_ will define the command arguments, then _functionB_ will implement the function.
+
+In our project we need two commands, let's set them up
+
+```javascript
+const argv = require("yargs");
+argv.command(
+  "joke",
+  "Fetching your joke",
+  (yargs) => {},
+  (argv) => {}
+).argv;
+
+argv.command(
+  "quote",
+  "Fetching a quote",
+  (yargs) => {},
+  (argv) => {}
+).argv;
+```
+
+Finally, let's pass in our functions to enable the jokes and quotes to be displayed.
```

Comment:
This part is confusing. Since this code snippet is similar to the one above. Could you just use the code snippet below and then explain about passing functions?

Reply:
yes, makes sense. I have removed the upper on and used the one with passed in functions to demonstrate

Reply:
Thanks, @katungi. Could you commit these changes?